### PR TITLE
PPTP-1363: Cancelify

### PR DESCRIPTION
**Pull request by lmcewanhmrc**
- Add cancel link to exit from amend journey, clears any answers
- If user amends different return, answers are first cleared

### CODE REVIEW

**Review by C-Thomp — APPROVED** [review]

File: app/controllers/amends/CheckYourAnswersController.scala [hunk 1]
```diff
@@ -122,11 +129,28 @@ class CheckYourAnswersController @Inject() (
...
+    (identify andThen getData andThen requireData).async {
+      implicit request =>
+        val pptReference: String = request.request.pptReference
+        val maybePeriodKey = request.userAnswers.get(AmendSelectedPeriodKey)
+        val futureNextPage = request.userAnswers
+          .reset
+          .save(cacheConnector.saveUserAnswerFunc(pptReference))
+          .map { _ =>
+            maybePeriodKey match {
+              case Some(periodKey) => routes.ViewReturnSummaryController.onPageLoad(periodKey)
```
Comment (mauro-fanton) [inline app/controllers/amends/CheckYourAnswersController.scala:150]: Should the Cancel redirect to a "are you sure you want to cancel the amend?" page or something similar? According to Adam and Rebecca this should be the case. But please ignore this message if you have plan to do this as part of a new task.

File: conf/messages.cy [hunk 2]
```diff
@@ -81,6 +81,7 @@ AmendsCheckYourAnswers.hiddenCell.action = [NEEDS WELSH] You cannot amend this f
 AmendsCheckYourAnswers.hiddenCell.newAnswer = [NEEDS WELSH] Not yet amended
 AmendsCheckYourAnswers.submit = [NEEDS WELSH] Submit amended return
 AmendsCheckYourAnswers.hiddenCell.changeLink = {0} {1}
+AmendsCheckYourAnswers.cancel.link.text = Cancel this amendment
```
Comment (mauro-fanton) [inline conf/messages.cy:84]: Should this be prefixes with [NEEDS WELSH]

**Review by mauro-fanton — APPROVED** [review]

File: app/models/UserAnswers.scala [hunk 3]
```diff
@@ -55,6 +55,30 @@ case class UserAnswers(
...
   }
 
+    /**
+    * Sets the answer (value) to the given question (settable), and returns a failed future  if something goes 
+    * wrong - unlike set() which returns a failed Try
+    * @param settable - question we're setting the answer / data for
+    * @param value - value to set
+    * @param cleanup - true to call the question's cleanup method
+    * @tparam A - type of 'value'
+    * @return Future of updated UserAnswers
```
Comment (liampan) [inline app/models/UserAnswers.scala:65]: Set returns a Try[UserAnswers], we then call a .get, do we want this to be a future? .get will throw exception

File: app/views/amends/CheckYourAnswersView.scala.html [hunk 4]
```diff
@@ -93,12 +95,11 @@ <h1 class="govuk-heading-xl">@messages("AmendsCheckYourAnswers.title", displayRe
...
-
-        @govukButton(
-            ButtonViewModel(messages("AmendsCheckYourAnswers.submit"))
-        )
-    }
+        @govukButton(ButtonViewModel(messages("AmendsCheckYourAnswers.submit")))
+        @paragraph(content = link(id = "cancel-link", text = messages("AmendsCheckYourAnswers.cancel.link.text"),
+            call = controllers.amends.routes.CheckYourAnswersController.cancel()))    }
+    
+    <br /><!--bottom margin-->
```
Comment (liampan) [inline app/views/amends/CheckYourAnswersView.scala.html:104]: does this work ? 🤔 
```suggestion
    <!--bottom margin-->
    <br> 
```

**Pull request merged**